### PR TITLE
feature/multivariate - step 3

### PR DESCRIPTION
**Pull request by pennfranc**
- Added multivariate support to `TorchForecastingModel` and both subclasses (`TCNModel` and `RNNModel`). Both input and predicted TimeSeries instances can now be multidimensional.
- Added `use_full_output_length` boolean argument to TorchForecastingModel.predict, which allows all outputs of the network to be used for the prediction. This argument must be set to True if multivariate TimeSeries are being predicted.
- Added multivariate support to `_TimeSeriesDataset1DSequential` and `_TimeSeriesDataset1DShifted`
- Added multivariate tests for `TCNModel` and `RNNModel`

### CODE REVIEW

**Comment by hrzn** [comment]
Could you update the darts intro notebook as well? I think at the beginning we explain that `TimeSeries` are wrappers around `pd.Series`. This can be updated :)

File: darts/tests/test_backtesting.py [hunk 1]
```diff
@@ -53,6 +60,38 @@ class BacktestingTestCase(unittest.TestCase):
...
+    def test_backtest_regression(self):
+        gaussian_series = gt(mean=2, length=50)
+        sine_series = st(length=50)
+        features = [gaussian_series + sine_series, gaussian_series]
+        features_multivariate = [(gaussian_series + sine_series).stack(gaussian_series), gaussian_series]
+        target = st(length=50)
+
+        # univariate feature test
+        pred = backtest_regression(features, target, StandardRegressionModel(15), pd.Timestamp('20000201'), 3)
+        self.assertEqual(r2_score(pred, target), 1.0)
```
Comment (hrzn) [inline darts/tests/test_backtesting.py:83]: I'm somewhat surprised we get *exactly* a R2 score of 1.0. I get that the model should learn to denoise perfectly, but I would expect the numeric procedure of fitting the model coefficients to converge only with some limited numeric accuracy.

File: darts/utils/missing_values.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def auto_fillna(ts: TimeSeries,
         A new TimeSeries with all missing values filled according to the rules above.
     """
 
-    ts_temp = ts.pd_series()
+    ts_temp = ts.pd_dataframe()
```
Comment (hrzn) [inline darts/utils/missing_values.py:67]: Line 57: change `pandas.Series` to `pandas.DataFrame`.

File: darts/tests/test_backtesting.py (hunk 1 above)
Comment (pennfranc) [reply darts/tests/test_backtesting.py:83]: So in this case we are using `sklearn.linear_model.LinearRegression`, which is a wrapper around `scipy.linalg.lstsq`, which is simply a solver for the ordinary least squares problem, which I assume is computed in a closed form (that has an ideal solution in this case). I guess this way the numeric accuracy should be pretty good?

Comment (hrzn) [reply darts/tests/test_backtesting.py:83]: Yes probably, you're right. Still it's quite neat to see that's exact in this case!

File: darts/models/torch_forecasting_model.py [hunk 3]
```diff
@@ -251,7 +276,7 @@ def fit(self,
             series: TimeSeries,
             val_series: Optional[TimeSeries] = None,
             verbose: bool = False,
-            component_index: Optional[int] = None) -> None:
+            target_indices: List[int] = []) -> None:
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:279]: > Also I didn't see where this parameter is used (incl. in `MultivariateForecastingModel`)

So far this parameter is only used in the `_TimeSeriesShiftedDataset` and `_TimeSeriesSequentialDataset` in `TorchForecastingModel`.

File: darts/models/torch_forecasting_model.py [hunk 4]
```diff
@@ -158,6 +168,16 @@ def __init__(self,
             Number of time steps to be output by the forecasting module.
         input_length
             Number of past time steps that are fed to the forecasting module.
+        input_size
+            The dimensionality of the TimeSeries instances that will be fed to the fit function.
+        output_size
+            The dimensionality of the output time series.
+        datetime_enhancements
+            A list of integers representing pd.Datetime attributes which will be included in the
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:176]: I actually removed this functionality, but forgot to clear the docstrings.

File: darts/models/tcn_model.py [hunk 5]
```diff
@@ -16,7 +16,18 @@
 
 class _ResidualBlock(nn.Module):
 
-    def __init__(self, num_filters, kernel_size, dilation_base, dropout, weight_norm, nr_blocks_below, num_layers):
+    def __init__(
+        self,
+        num_filters,
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/tcn_model.py]: done

File: darts/models/torch_forecasting_model.py [hunk 6]
```diff
@@ -39,7 +40,8 @@ class _TimeSeriesDataset1DSequential(Dataset):
     def __init__(self,
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:40]: done

File: darts/models/torch_forecasting_model.py [hunk 7]
```diff
@@ -78,15 +81,16 @@ def __getitem__(self, index):
         idx = index % (self.len_series - self.data_length - self.target_length + 1)
         data = self.series_values[idx:idx + self.data_length]
         target = self.series_values[idx + self.data_length:idx + self.data_length + self.target_length]
-        return torch.from_numpy(data).float().unsqueeze(1), torch.from_numpy(target).float().unsqueeze(1)
+        return torch.from_numpy(data).float(), torch.from_numpy(target[:, self.target_indices]).float()
 
 
 class _TimeSeriesDataset1DShifted(torch.utils.data.Dataset):
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py]: done

File: darts/models/torch_forecasting_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py]: done

File: darts/models/torch_forecasting_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py]: done

File: darts/models/torch_forecasting_model.py [hunk 8]
```diff
@@ -309,20 +337,41 @@ def fit(self,
             tb_writer.flush()
             tb_writer.close()
 
-    def predict(self, n: int) -> TimeSeries:
+    def predict(self, n: int, use_full_output_length: bool = False) -> TimeSeries:
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:331]: done

File: darts/utils/missing_values.py (hunk 2 above)
Comment (pennfranc) [reply darts/utils/missing_values.py:67]: done

File: darts/models/torch_forecasting_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hrzn) [reply darts/models/torch_forecasting_model.py:331]: Thanks it's better, however I think it can still be improved. "whether or not the full output sequence of the model prediction should be used to produce the output of this function" doesn't say too much.
If it's used, how is it used? And if it's not used, how is the output obtained?

File: darts/models/torch_forecasting_model.py [hunk 9]
```diff
@@ -309,20 +337,41 @@ def fit(self,
...
+        # iterate through predicting output and consuming it again until enough predictions are created
         test_out = []
         self.model.eval()
-        for i in range(n):
-            out = self.model(pred_in)
-            pred_in = pred_in.roll(-1, 1)
-            pred_in[:, -1, :] = out[:, self.first_prediction_index]
-            test_out.append(out.cpu().detach().numpy()[0, self.first_prediction_index])
+        if (use_full_output_length):
+            num_iterations = math.ceil(n / self.output_length)
```
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py]: done

File: darts/models/torch_forecasting_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pennfranc) [reply darts/models/torch_forecasting_model.py:331]: The more detailed description of the mechanism: "If `use_full_output_length` is set to `False`, the model will perform..." (line 335) is included in the block of text above the argument descriptions. Do you think I should move it to the description of `use_full_output_length`, or is it also not detailed enough?

[the remark this answers is not in the document]
Comment (hrzn) [reply darts/models/torch_forecasting_model.py:331]: Ohhh sorry I missed this. That's all good then :)
